src/_pytest/hookspec.py

```diff
@@ -597,6 +597,10 @@ def pytest_warning_captured(warning_message, when, item):
         in a future release.
 
         The item being executed if ``when`` is ``"runtest"``, otherwise ``None``.
+
+    :param types.FrameType frame:
```

Comment:
Hi @phloose, sorry for the delay!

That's definitely an interesting approach! 

`call_historic` here means that this hook will be called even for plugins registered after this call was made, if they implement the hook (see [https://pluggy.readthedocs.io/en/latest/index.html#historic-hooks](https://pluggy.readthedocs.io/en/latest/index.html#historic-hooks)).

Problem is that, to implement that, `call_historic` needs to keep the parameters passed to it around *forever*, which means that all locals on that frame will also be kept around, meaning a huge memory leak. 😕

Reply:
Ok sounds reasonable... Didn't have such a deep insight into pluggy and didnt thought of a memory leak regarding the locals from the frame and `call_historic`. I just thought of a way to not only match against strings/testoutput/linenumbers when the exact warning location is important (described here: https://github.com/pytest-dev/pytest/pull/5984#issuecomment-543776819). So should i revert the recent changes regarding the frame thing to maybe 3237e05 ?

Reply:
Sounds reasonable, except the last commit is "Add test for MarkGenerator warning location" which I guess should be kept. But I think you can squash all the commits together anyway for the final merge. 👍

Reply:
Did an interactive rebase and removed the frame related commits. I am still questioning how to check for the exact location of a warning, if linenumbers of a warning are too strict as outlined in https://github.com/pytest-dev/pytest/pull/5984#issuecomment-543776819...

Reply:
Yeah it is not clear to me as well, but having the test failures at least alerts us to the problem, which is better than letting the warning point to an unrelated location after a refactoring. 😕

Reply:
Hmm ok. I am wondering how for a large module like `_pytest.config.__init__.py` (`config._preparse` and `config.PytestPluginManager.import_plugin`) this can be helpful at all if you only check that the error comes from this module and not in which context, because then the stacklevel changes cannot be precisely checked in the test - it could be anywhere in that module... But yeah since checking against frame is no option for obvious reasons and linenumbers are too strict with respect to refactorings this is maybe the only way until now.

Reply:
Maybe it could pass the frame only when under test / via some config?
Otherwise it might also be not a big problem after all keeping the frame(s) around?  (there should not be many warning after all)

Reply:
Also not handling the frame for historic calls might be an option, i.e. call it (passing a reference), but set it to `None` afterwards.
Also considering if `pytest_warning_captured` really needs to be historic might be worth it (IIRC there some issue about leaks with warnings already, which might be related).

Reply:
Also we could of course not pass the frame as is, but copies of values we can use for matching then (e.g. `co_name`).

Reply:
> Also we could of course not pass the frame as is, but copies of values we can use for matching then (e.g. `co_name`).

This sounds feasible to me 🤔 worth rethinking the whole thing ... Unfortunatly i threw away the commits with the frame related stuff. Maybe i am lucky looking in the reflog. Then we would need to find a way to distinguish clearly between names for modules and functions (`co_name` evaluates to the string `<module>` instead of the module's name when the context is a module)

Reply:
@phloose 
Yeah, another case of having to force-push here before merging.. :/
But Github keeps track of this, see the comments:

> phloose force-pushed the phloose:acceptance-tests_pytest-dev/pytest/#4445  branch from 591328c to c489af3 10 days ago
> phloose force-pushed the phloose:acceptance-tests_pytest-dev/pytest/#4445  branch from ac0df83 to 46663af 18 hours ago

So you can just `git reset --hard …` with it (create a backup if you're not experienced with resetting.. ;))

I wonder if you could just configure the warnings format with the (hopefully single) outer run to include more/necessary information then in the output?